# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/painting/nsDisplayList.cpp
# Commit: 13f70dfa4e1d
# Full Hash: 13f70dfa4e1d1d2c0f00b851ca655529f624fddd
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-16 09:17:48
# Regressor Bug: 1540737
# File Overlap Count: 2
# Description:
#   Bug 1540737 - Allow filter painting to take a callback for painting children. r=miko
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114670
# ==============================================================================

diff -r 10dacf71ccbe -r 13f70dfa4e1d layout/painting/nsDisplayList.cpp
--- a/layout/painting/nsDisplayList.cpp	Sun May 16 02:28:40 2021 +0300
+++ b/layout/painting/nsDisplayList.cpp	Sun May 16 03:55:29 2021 +0000
@@ -9906,12 +9906,37 @@
 
 void nsDisplayFilters::PaintAsLayer(nsDisplayListBuilder* aBuilder,
                                     gfxContext* aCtx, LayerManager* aManager) {
+  PaintWithContentsPaintCallback(aBuilder, aCtx, [&](gfxContext* aContext) {
+    BasicLayerManager* basic = aManager->AsBasicLayerManager();
+    RefPtr<gfxContext> oldCtx = basic->GetTarget();
+    basic->SetTarget(aContext);
+
+    aManager->EndTransaction(FrameLayerBuilder::DrawPaintedLayer, aBuilder);
+    basic->SetTarget(oldCtx);
+  });
+}
+
+void nsDisplayFilters::PaintWithContentsPaintCallback(
+    nsDisplayListBuilder* aBuilder, gfxContext* aCtx,
+    const std::function<void(gfxContext* aContext)>& aPaintChildren) {
   imgDrawingParams imgParams(aBuilder->GetImageDecodeFlags());
   nsRect borderArea = nsRect(ToReferenceFrame(), mFrame->GetSize());
   SVGIntegrationUtils::PaintFramesParams params(*aCtx, mFrame, GetPaintRect(),
-                                                borderArea, aBuilder, aManager,
+                                                borderArea, aBuilder, nullptr,
                                                 mHandleOpacity, imgParams);
-  SVGIntegrationUtils::PaintFilter(params);
+
+  gfxPoint userSpaceToFrameSpaceOffset =
+      SVGIntegrationUtils::GetOffsetToUserSpaceInDevPx(mFrame, params);
+
+  SVGIntegrationUtils::PaintFilter(
+      params,
+      [&](gfxContext& aContext, nsIFrame* aTarget, const gfxMatrix& aTransform,
+          const nsIntRect* aDirtyRect, image::imgDrawingParams& aImgParams) {
+        gfxContextMatrixAutoSaveRestore autoSR(&aContext);
+        aContext.SetMatrixDouble(aContext.CurrentMatrixDouble().PreTranslate(
+            -userSpaceToFrameSpaceOffset));
+        aPaintChildren(&aContext);
+      });
   nsDisplayFiltersGeometry::UpdateDrawResult(this, imgParams.result);
 }
 